docs/articles/documentation/recipes/best-practices/best-practices.md

```diff
@@ -0,0 +1,223 @@
+
+## File Structure
+
+Follow these guidelines to keep your test structure manageable and "clean":
+
+* If you use [page model](../../guides/concepts/page-model.md) for your tests (which is recommended), put all the page model files into one directory. If your application is divisible into logical chunks (components or subsystems), split up the associated page model objects into separate files.
+
+* Use the page model to store Selectors and compound actions that are common in your app. For instance, put all actions that are necessary to perform a login into one reusable function.
+
+> You can find an example of a page model in the [testcafe-examples](https://github.com/DevExpress/testcafe-examples/tree/master/examples/use-page-model) repository.
+
+* Use the `.testcaferc.json` configuration file to fine-tune your tests, which can be useful in CI/CD systems. This file must be located in the root directory of the project. For more information, read the [Configuration File](../../reference/configuration-file.md) article.
+
+* Define one `fixture` in every test file. While it's technically possible to define multiple, this is not recommended.
+
+* TestCafe tests are purely functional. As such, they should not be concerned about the implementation details and it is best to isolate them from production code. Keep your test files in a separate directory. You can name this directory appropriately (for instance, `tests`).
+
+* In your test folder, group your tests into subfolders. These groups cover different subsystems of your application with tests.
+
+* Any reused data (for example, large sets of expected values or forms inputs) is better stored in a directory of its own. Consider a descriptive folder name (for instance, `data`).
+
+With all the suggestions applied, your project's file structure might look like this:
+
+```sh
+.
+├── .testcaferc.json
+└── tests
+    ├── |- test_group1/
+    │   └── |-test1.js
+    │       └── |-test2.js
+    ├── |- test_group2/
+    │   └── |-test1.js
+    │       └── |-test2.js
+    ├── |- page_model/
+    │   └── |- file1.js
+    │       └── |- file2.js
+    ├── |- helpers/
+    │   └── |- helper1.js
+    │       └── |- helper2.js
+    └── |-data
+```
+
+## Cleanup With beforeEach / State Management
+
+State management is an integral and important part of web testing. When your tests run, there are inevitably leftovers- database or local storage records, cache, or cookies. Your tests might require setup to run - like adding records to the database.
+
+One common strategy is to set up with the previous test's `after` and `afterEach` hooks. While good for cleanup, these hooks create mutual dependence between your tests when used to set up for the following test. The success rate of a test is then influenced by a preceding one, which is not desirable.
```

Comment:
I can't believe I'm saying this, but I couldn't get what `with` means here at first. I'm also used to writing "with" as a substitution for "using", but `set up with hooks` really puzzles. Now I understand Dirk.

We can say *to place the setup actions in... hooks* or *to use... hooks for setup* or something like that.

Applies to some more occurrences below.

Reply:
> One common strategy is to place the setup actions in the previous test's `after` and `afterEach` hooks.

> Plus, if the `before`/`beforeEach` setup is unsuccessful, the associated test does not run, which saves you time. If `after`/`afterEach` setup yields an error, the following test runs and probably fails due to the unsuccessful setup.